### PR TITLE
ref(fairmapper):  multi-pass PK resolution + cleanup in ResolveMissingPkPostProcessor

### PR DESCRIPTION
**Pull request by jhhaanstra**
Addresses: https://github.com/molgenis/molgenis-emx2/issues/6629

### What are the main changes you did
- Fixed a bug in `ResolveMissingPkPostProcessor` where a reference could stay unresolved forever, depending on table processing order. Tables are resolved in a fixed order, so if Orders (which references Products) is processed before Products has filled in its own composite key, Orders.product would be left empty — there was no second chance to fill it in later.

- The fix: instead of doing one pass over all tables, the processor now keeps repeating passes (up to 10) until a full pass makes no more changes. So a table processed early can still pick up values that only became available once a later table resolved.

### How to test

- Run `./gradlew :backend:molgenis-emx2-fairmapper:test`, ResolveMissingPkPostProcessorTest covers:
  - single and array references, 
  - composite/mutual-key references
  - two regression tests specifically for the ordering bug: `givenResolving_whenTableUpdatedAfterHandledByResolver_thenResolveAgain` and its array-reference counterpart, both of which set up tables so the dependency can only resolve on a second pass.

### Checklist

- [ ] checked that code complies with the [dev guidelines](docs/molgenis/dev_guidelines.md)
- [ ] frontend code follows good semantic HTML practices and meets accessibility guidelines [Accessibility guide](docs/molgenis/dev_accessibility.md)
- [ ] updated docs in case of new feature
- [ ] added/updated tests
- [ ] added/updated testplan to include a test for this fix, including ref to bug using # notation

By creating this pull request I have agreed with the [contributor terms](https://github.com/molgenis/molgenis-emx2/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by mswertz** [comment]
in MOLGENIS backend we do a dependency ordering. Now, for indirect circular dependencies that is not easy. That is why MOLGENIS solves references at the end. Long story short, did you consider dependency order in this? To reduce loops? (I might need more carefully review the code).

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 1]
```diff
@@ -1,148 +1,211 @@
...
-          tableName,
-          (iterator, source) ->
-              iterator.forEachRemaining(row -> resolveRow(tableStore, row, referenceColumns)));
+    int nrIterations = 0;
+
+    while (nrIterations < MAX_NR_ITERATIONS) {
+      changesMade = false;
+      nrIterations++;
+
+      for (String tableName : schema.getTableNames()) {
```
Comment (mswertz) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:68]: so here you could get dependency ordered (or remind me, is this already dependency ordered?)

**Review by mswertz — APPROVED** [review]

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 2]
```diff
@@ -1,148 +1,211 @@
...
+ *
+ *   Customer
+ *     customerNumber  (key, part 1)
+ *     order       ->  Order.orderNumber   (key, part 2)
+ * </pre>
+ *
+ * <p>To fill in {@code customer} on the {@code Order} row, we need {@code Customer.order} to
+ * already be filled in. But {@code Customer.order} can only be filled in using this very {@code
+ * Order} row. Waiting for {@code Customer} to resolve first would never work, since it's waiting on
+ * us too. So instead we just copy {@code Order}'s own {@code orderNumber} straight into {@code
```
Comment (svandenhoek) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:39]: "waiting on us too" -> might wanna rephrase this.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 3]
```diff
@@ -1,148 +1,211 @@
...
+ * Customer.order}. See {@link #pointsBackAtOwnTable} and {@link #completeMutualKey} for where this
+ * happens.
+ *
+ * <p>Tables are resolved in schema order, so a table processed early may need a value that only
+ * becomes available once a later table is resolved (or gets this kind of fix-up). {@link #process}
+ * therefore keeps repeating passes over every table until a pass makes no more changes, or a
+ * maximum number of iterations is reached.
```
Comment (svandenhoek) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:46]: Maybe we should use [TableSort](https://github.com/molgenis/molgenis-emx2/blob/master/backend/molgenis-emx2/src/main/java/org/molgenis/emx2/utils/TableSort.java) here? Might not fix this issue completely, but it orders the tables where "dependency come from foreign key and from inheritance". So even if we still need more than 1 pass, it should at least reduce this number.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 4]
```diff
@@ -1,148 +1,211 @@
...
+ */
 public class ResolveMissingPkPostProcessor implements PostProcessor {
 
   private static final Logger logger = LoggerFactory.getLogger(ResolveMissingPkPostProcessor.class);
-  private static final String SUBJECT_VARIABLE = "_subject_";
+  private static final String SUBJECT = "_subject_";
+  private static final int MAX_NR_ITERATIONS = 10;
```
Comment (svandenhoek) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:51]: Seems quite arbitrary? After implementing usage of [TableSort](https://github.com/molgenis/molgenis-emx2/blob/master/backend/molgenis-emx2/src/main/java/org/molgenis/emx2/utils/TableSort.java), I'm curious how this number would change with the current test dataset.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 1 above)
Comment (svandenhoek) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:68]: I don't think so (IDEA claims [TableSort](https://github.com/molgenis/molgenis-emx2/blob/master/backend/molgenis-emx2/src/main/java/org/molgenis/emx2/utils/TableSort.java) is only used in `SqlSchema).`

EDIT: In conversation view it does not show up as a reply to comment of @mswertz. See code changes view to see it as a reply.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 5]
```diff
@@ -1,148 +1,211 @@
...
+        changesMade = true;
+      }
+    }
   }
 
-  private void resolveArray(TableStore tableStore, Row row, Column column) {
-    if (!row.containsName(SUBJECT_VARIABLE + column.getName())) {
+  private void resolveArrayReference(TableStore tableStore, Row row, Column column) {
+    String field = subjectField(column);
+    if (!row.notNull(field)) {
```
Comment (svandenhoek) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:128]: See above.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 6]
```diff
@@ -1,148 +1,211 @@
...
       }
     }
   }
 
-  private void resolveSingular(TableStore tableStore, Row row, Column column) {
-    String subject = row.getString(SUBJECT_VARIABLE + column.getName());
-    if (subject == null) {
+  private void resolveSingleReference(TableStore tableStore, Row row, Column column) {
+    String field = subjectField(column);
+    if (!row.notNull(field)) {
```
Comment (svandenhoek) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:100]: Any reason to not use `row.isNull()`? If this would result in different behaviour, I think that would need to be fixed there?

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 7]
```diff
@@ -1,148 +1,211 @@
...
+  private static Optional<Object> readAvailableValue(Reference reference, Row referencedRow) {
+    if (referencedRow.notNull(reference.getReferencedColumnName())) {
+      return Optional.of(referencedRow.getValueMap().get(reference.getReferencedColumnName()));
+    }
+    return Optional.empty();
+  }
+
+  private static String subjectField(Column column) {
+    return SUBJECT + column.getName();
   }
```
Comment (svandenhoek) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java]: Maybe these two can be merged into a single method where the only difference is that the non-array version at 1 step simply returns a list with 1 item?

File: backend/molgenis-emx2-fairmapper/src/test/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessorTest.java [hunk 8]
```diff
@@ -169,59 +193,219 @@ void setup() {
...
+    // after Orders has already been handled. Orders should be resolved again once that new
+    // information becomes available, instead of keeping the incomplete composite key array it
+    // read on its first pass.
+    Row order = tableStore.readTable("Orders").iterator().next();
+    assertArrayEquals(new String[] {"product-1"}, order.getStringArray("products.id"));
+    assertArrayEquals(new String[] {"supplier-1"}, order.getStringArray("products.supplier"));
+  }
+
   @Nested
   class BackReferenceTest {
```
Comment (svandenhoek) [inline backend/molgenis-emx2-fairmapper/src/test/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessorTest.java:410]: Consider adding a more complicated test where there is a table in between as well:
```
Order
  orderNumber   (key)
  customerGroup  ->  CustomerGroup

CustomerGroup
  groupNumber (key)
  members -> Customer
 
Customer
  customerNumber  (key, part 1)
  order       ->  Order.orderNumber   (key, part 2)
```

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 9]
```diff
@@ -1,148 +1,211 @@
...
         }
       }
-      row.setRefArray(reference.getColumnName(), values.toArray());
+
+      if (allResolved) {
+        row.setRefArray(reference.getColumnName(), values.toArray());
+        changesMade = true;
+      }
     }
   }
```
Comment (svandenhoek) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:171]: A lot of duplicate code. I would suggest an alternative implementation instead:
```java
  private void resolveRow(TableStore tableStore, Row row, List<Column> referenceColumns) {
    for (Column column : referenceColumns) {
      String field = subjectField(column);
      if (!row.notNull(field)) {
        continue;
      }

      if (column.isArray()) {
        resolveReference(
            row,
            column,
            Arrays.stream(row.getStringArray(field))
                .map(
                    subject ->
                        TableStoreUtils.getRowForSubject(tableStore, column.getRefTable(), subject))
                .toList(),
            (reference, values) -> row.setRefArray(reference.getColumnName(), values.toArray()));
      } else {
        resolveReference(
            row,
            column,
            List.of(
                TableStoreUtils.getRowForSubject(
                    tableStore, column.getRefTable(), row.getString(field))),
            (reference, values) -> row.set(reference.getColumnName(), values.getFirst()));
      }
    }
  }

  private void resolveReference(
      Row row,
      Column column,
      List<Row> referencedRows,
      BiConsumer<Reference, List<Object>> rowSetter) {
    for (Reference reference : column.getReferences()) {
      // The field might already hold an empty array left over from extraction, not a real
      // result. Only treat it as resolved once it has as many values as there are rows.
      if (isFullyResolvedArray(row, reference, referencedRows.size())) {
        continue;
      }

      List<Object> values = new ArrayList<>();
      boolean allResolved = true;
      for (Row referencedRow : referencedRows) {
        Optional<Object> value = readAvailableValue(reference, referencedRow);
        if (value.isEmpty() && pointsBackAtOwnTable(column, reference)) {
          // These two rows depend on each other (see the class doc example): the referenced row
          // can't resolve this on its own, so we write the value into both rows right here.
          value = Optional.of(completeMutualKey(row, referencedRow, reference));
        }

        if (value.isPresent()) {
          values.add(value.get());
        } else {
          // Not resolvable yet. Leave the array untouched below so a later pass can retry --
          // writing a partial result now would make it look already done. Keep going through
          // the rest of the rows anyway, since other referenced rows here may still need their
          // own missing key part completed via completeMutualKey.
          allResolved = false;
        }
      }

      if (allResolved) {
        rowSetter.accept(reference, values);
        changesMade = true;
      }
    }
  }
```

Comment (svandenhoek) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:171]: Note that this snippet does not have the suggested `row.isNull()` suggestion implemented.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 7 above)
Comment (svandenhoek) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:183]: `TableStoreUtils` also defines the `SUBJECT`, so I would suggest moving this method to there or create some other solution so that the constants & generic methods aren't repeated in multiple locations (and `ResolveRefbackPostProcessor` also has this constant for that matter).

**Review by svandenhoek — CHANGES_REQUESTED** [review]

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 10]
```diff
@@ -1,148 +1,212 @@
...
-              .toList();
-
-      tableStore.processTable(
-          tableName,
-          (iterator, source) ->
-              iterator.forEachRemaining(row -> resolveRow(tableStore, row, referenceColumns)));
+    int nrIterations = 0;
+
+    while (nrIterations < MAX_NR_ITERATIONS) {
+      changesMade = false;
```
Comment (aikevd) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java]: This var is not used/written to?

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 4 above)
Comment (jhhaanstra) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:51]: There didn't go a lot of thought into this number:
- It's a safeguard to prevent infinite loops
- If you haven't resolved after 10 loops, you will **probably** won't resolve at all
- I think the overhead of 5 vs 10 is negligible for these use cases.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 3 above)
Comment (jhhaanstra) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:46]: Added table sorting

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 1 above)
Comment (jhhaanstra) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:68]: I added the table sort as sander suggested.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 11]
```diff
@@ -1,148 +1,212 @@
...
+      nrIterations++;
+
+      for (String tableName : schema.getTableNames()) {
+        List<Column> referenceColumns = getReferenceColumnsForTable(tableName);
```
Comment (aikevd) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:69]:
```suggestion
        List<Column> referenceColumns = getReferenceColumnsForTable(tableName);
        if (referenceColumns.isEmpty()) continue;
```

Skip table iteration when no referenceColumns

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 12]
```diff
@@ -1,148 +1,212 @@
...
-    if (subject == null) {
+  private void resolveSingleReference(TableStore tableStore, Row row, Column column) {
+    String field = subjectField(column);
+    if (!row.notNull(field)) {
       return;
     }
 
-    Row referringRow = getRowForSubject(tableStore, column.getRefTableName(), subject);
+    String subject = row.getString(field);
+    Row referencedRow = TableStoreUtils.getRowForSubject(tableStore, column.getRefTable(), subject);
```
Comment (aikevd) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:105]: Did you consider creating a lookup index (`Subject -> Row`) instead of doing a new scan for each reference, to allow better performance scaling?

Comment (aikevd) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:105]: E.g.
```java
final class SubjectIndex {

  private final TableStore tableStore;
  private final Map<String, Map<String, Row>> cache = new HashMap<>();

  SubjectIndex(TableStore tableStore) {
    this.tableStore = tableStore;
  }

  Row get(TableMetadata table, String subject) {
    for (TableMetadata candidate : table.getInheritanceTree()) {
      Row row = indexFor(candidate.getTableName()).get(subject);
      if (row != null) {
        return row;
      }
    }
    throw new MolgenisException(
        "Referencing non-existing row for table: " + table.getTableName()
            + ", for subject: " + subject);
  }

  private Map<String, Row> indexFor(String tableName) {
    return cache.computeIfAbsent(tableName, this::buildIndex);
  }

  private Map<String, Row> buildIndex(String tableName) {
    Map<String, Row> bySubject = new HashMap<>();
    for (Row row : tableStore.readTable(tableName)) {
      String subject = row.getString(SUBJECT_NAME);
      if (subject != null) {
        bySubject.putIfAbsent(subject, row);
      }
    }
    return bySubject;
  }
}
```

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 6 above)
Comment (aikevd) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:100]: Interesting point from Sander, which exposes the detail that `Row.notNull()` also counts an empty string `""` as 'null' (i.e. this check will `return`).

That means that if the harvested data contains empty strings as references, this check would not work.
I think you can best estimate yourself wheteher that is a real or mostly theoretical problem @jhhaanstra

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 13]
```diff
@@ -1,148 +1,212 @@
...
-        if (isCircularReference(column, reference)) {
-          Object value = row.getValueMap().get(reference.getTargetColumn());
-          updateReferringRow(row, reference, referringRow, value);
-          row.set(reference.getColumnName(), value);
-        }
+      Optional<Object> value = readAvailableValue(reference, referencedRow);
+      if (value.isEmpty() && pointsBackAtOwnTable(column, reference)) {
+        // These two rows depend on each other (see the class doc example): the referenced row
+        // can't resolve this on its own, so we write the value into both rows right here.
+        value = Optional.of(completeMutualKey(row, referencedRow, reference));
```
Comment (aikevd) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:116]: While trying to get my head around this, Claude pointed out a (theoretical) scenario where this solution is too eager:
```
Orders:
  order-1  (currently being resolved, points at supplier-1)
  order-7  (a different, unrelated order)

Suppliers:
  supplier-1  →  really belongs to order-7  (its RDF data says so directly)

Code flow:
1. order-1.supplier needs supplier-1's full key: (id, order).
2. supplier-1.id is already known: "supplier-1". Fine, no shortcut needed there.
3. supplier-1.order is still empty at this exact moment — the row for Suppliers hasn't had its own turn processed yet.
4. The code checks: does Suppliers.order point back at Orders? Yes → pointsBackAtOwnTable is true.
5. Shortcut fires: it takes order-1's own id and force-writes it into supplier-1.order.

But supplier-1 was never supposed to belong to order-1. It belongs to order-7. The code just guessed wrong, using the "shortcut" meant only for genuine deadlocks.
```

I see that this logic already existed (outside this PR) but it may need to be flagged for a fix?

Comment (aikevd) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:116]: Note: Claude suggests to e.g. delay this 'shortcut resolution' until the normal resolutions are exhausted, e.g.:
```java
boolean allowFabrication = false;
while (nrIterations < MAX_NR_ITERATIONS) {
  changesMade = false;
  ... passes ...
  if (!changesMade) {
    if (allowFabrication) break;      // stalled even with fabrication → truly done
    allowFabrication = true;          // stalled normally → now allow the last resort
  }
}
```

I'm not sure if this is the definitive solution as it doesn't make the algorithm simpler ;-), but perhaps it helps shaping a solution

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 14]
```diff
@@ -1,148 +1,212 @@
...
       }
 
       List<Object> values = new ArrayList<>();
-      for (Row referringRow : rows) {
-        if (referringRow.notNull(reference.getReferencedColumnName())) {
-          values.add(referringRow.getValueMap().get(reference.getReferencedColumnName()));
+      boolean allResolved = true;
+      for (Row referencedRow : referencedRows) {
+        Optional<Object> value = readAvailableValue(reference, referencedRow);
+        if (value.isEmpty() && pointsBackAtOwnTable(column, reference)) {
```
Comment (aikevd) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:151]: Minor: Perhaps copy the clarifying comments from line 114 here too?

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 9 above)
Comment (aikevd) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:171]: I agree there is duplication but don't have a strong opinion whether that should be resolved; i think it's acceptable in current state (well readable and no high maintenance risk), but if you can dry it up while keeping it readable I agree that can be a further improvement.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 7 above)
Comment (aikevd) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:183]: I agree: the `_subject_` field naming convention is now spread across multiple parts in the pipeline (`SparqlVariableUtil, SparqlSelectRdfTransformer, ResolveMissingPkPostProcessor`) which poses a maintainability risk.

Perhaps you can move `subjectField` to `SparqlVariableUtill` which already contains subject-string logic? E.g. next to the existing `subjectVariable` method:
```java
// SPARQL query side: encodes, returns a Variable
public static Variable subjectVariable(Column column) { ... unchanged ... }

// Row storage side: no encoding needed, returns a String
public static String subjectField(Column column) {
  return SUBJECT_NAME + column.getName();
}
```

**Review by aikevd — CHANGES_REQUESTED** [review]
Nice work! I had to dig deep to find some meaningful remarks but managed to find some ;) :)

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 9 above)
Comment (jhhaanstra) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:171]: I gave this some more thought and also discussed this with @aikevd I decided to accept the duplicate code. The cognitive load of this problem is already pretty large and we both found that the deduplicated code made it less readable. So we prefer to handle the rule of three here, instead of immediately refactoring.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 12 above)
Comment (jhhaanstra) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:105]: Discussed and decided that what the pro's of using a cache doesn't hold up to its cons:
- more complexity for an already complex problem
- Doesn't reduce heap, but only saves some lookup time, which isn't a problem that we're running into

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 6 above)
Comment (svandenhoek) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:100]: @aikevd interesting. Maybe we should differentiate between these as `notNull` vs `notEmpty`? Current naming scheme sounds confusing.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 9 above)
Comment (svandenhoek) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:171]: In that case consider at least to reduce duplicate code by creating methods/moving some code. Some examples:

---
First few lines in both is identical:
```java
    String field = subjectField(column);
    if (!row.notNull(field)) {
      return;
    }
```
This could be moved to a method that calls either of them:
```java
    for (Column column : referenceColumns) {
      String field = subjectField(column);
      if (!row.notNull(field)) {
        continue;
      }

      if (column.isArray()) {
        resolveArrayReference(tableStore, row, column, field);
      } else {
        resolveSingleReference(tableStore, row, column, field);
      }
    }
```

---
The following code is very similar:
```java
String subject = row.getString(field);
Row referencedRow = TableStoreUtils.getRowForSubject(tableStore, column.getRefTable(), subject);
```
```java
TableMetadata refTable = column.getRefTable();
List<Row> referencedRows =
    Arrays.stream(row.getStringArray(field))
        .map(subject -> TableStoreUtils.getRowForSubject(tableStore, refTable, subject))
        .toList();
```
These could be converted into a single method both could call:
```java
  private List<Row> getReferencedRows(TableStore tableStore, Column column, String... values) {
    return Arrays.stream(values)
        .map(subject -> TableStoreUtils.getRowForSubject(tableStore, column.getRefTable(), subject))
        .toList();
  }
```
```java
Row referencedRow = getReferencedRows(tableStore, column, row.getString(field)).getFirst();
```
```java
List<Row> referencedRows = getReferencedRows(tableStore, column, row.getStringArray(field));
```

Comment (svandenhoek) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:171]: Maybe second example might not be the most efficient (useless creation of a list for one of them), but if cognitive load is the main issue then moving code to shared functions where possible seems exactly the thing we want to do?

**Review by svandenhoek — CHANGES_REQUESTED** [review]
Code looks fine, though I do suggest at least some refactoring of duplicate code. Perhaps my initial suggestion was a bit over-engineered, but there's definitely some ground to be gained by combining some of it.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 6 above)
Comment (aikevd) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:100]: Yes, that should definitely help understanding control flows around Row.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 9 above)
Comment (jhhaanstra) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:169]: Second one makes things more difficult, but the first one is a nice suggestion

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 13 above)
Comment (jhhaanstra) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java]: Good catch, this is an edge case where the ordering of the rows determines the value, but should be deterministic instead. I have been able to reproduce this in a test and created an issue to fix it. This is already broken on master and doesn't make or break the CSVW feature. So I chose to not fix it here, but via the other issue instead: https://github.com/molgenis/GCC/issues/3503

Comment (aikevd) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java]: Excellent 👍

**Review by svandenhoek — APPROVED** [review]

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 15]
```diff
@@ -1,148 +1,208 @@
...
-        if (isCircularReference(column, reference)) {
-          Object value = row.getValueMap().get(reference.getTargetColumn());
-          updateReferringRow(row, reference, referringRow, value);
-          row.set(reference.getColumnName(), value);
-        }
+      Optional<Object> value = readAvailableValue(reference, referencedRow);
+      if (value.isEmpty() && pointsBackAtOwnTable(column, reference)) {
+        // These two rows depend on each other (see the class doc example): the referenced row
+        // can't resolve this on its own, so we write the value into both rows right here.
+        value = Optional.of(completeMutualKey(row, referencedRow, reference));
```
Comment (aikevd) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java]: Since you use `isPresent` below, I think you meant this?
```suggestion
        value = Optional.ofNullable(completeMutualKey(row, referencedRow, reference));
```

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 16]
```diff
@@ -1,148 +1,208 @@
...
-              .filter(Column::isReference)
-              .toList();
-
-      tableStore.processTable(
-          tableName,
-          (iterator, source) ->
-              iterator.forEachRemaining(row -> resolveRow(tableStore, row, referenceColumns)));
+    int nrIterations = 0;
+
+    while (nrIterations < MAX_NR_ITERATIONS) {
```
Comment (aikevd) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:64]: Just checking: the code does not log any warnings when `MAX_NR_ITERATIONS` is exceeded. Do we want to flag this scenario?

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java [hunk 17]
```diff
@@ -1,148 +1,208 @@
...
+ * maximum number of iterations is reached.
+ */
 public class ResolveMissingPkPostProcessor implements PostProcessor {
 
   private static final Logger logger = LoggerFactory.getLogger(ResolveMissingPkPostProcessor.class);
-  private static final String SUBJECT_VARIABLE = "_subject_";
+  private static final int MAX_NR_ITERATIONS = 10;
 
   private final SchemaMetadata schema;
+  private boolean changesMade;
```
Comment (aikevd) [inline backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:54]: Checking: This class is instantiated in `DCATPostProcessor`. Could there be concurrency issues through the `changesMade` flag when e.g. the PostProcessor is used to handle multiple harvests in parallel?

**Review by aikevd — APPROVED** [review]
💪 🥳 
I have a few remaining remarks, none of which I estimate blocking.

File: backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java (hunk 17 above)
Comment (jhhaanstra) [reply backend/molgenis-emx2-fairmapper/src/main/java/org/molgenis/emx2/fairmapper/postprocessing/ResolveMissingPkPostProcessor.java:54]: Good point, and yes, concurrency is not taken into account when creating this class. We'd have to move the changesMade to be passed around more to make it support parallelism, but that's out of scope for now.
